packages/mui-codemod/src/v5.0.0/joy-rename-components-to-slots.js

```diff
@@ -0,0 +1,56 @@
+function transformComponentsProp(attributeNode) {
+  attributeNode.name.name = 'slots';
+
+  const valueExpression = attributeNode.value.expression;
+  if (valueExpression?.type !== 'ObjectExpression') {
+    return;
+  }
+
+  valueExpression.properties.forEach((property) => {
+    property.key.name = property.key.name[0].toLowerCase() + property.key.name.slice(1);
+
+    if (property.shorthand) {
+      property.shorthand = false;
+    }
+  });
+}
+
+function transformComponentsPropsProp(attributeNode) {
+  attributeNode.name.name = 'slotProps';
+}
+
+/**
+ * @param {import('jscodeshift').FileInfo} file
+ * @param {import('jscodeshift').API} api
+ */
+export default function transformer(file, api, options) {
+  const j = api.jscodeshift;
+  const root = j(file.source);
+  const printOptions = options.printOptions;
+
+  const transformed = root.findJSXElements().forEach((path) => {
```

Comment:
@hbjORbj To make codemod specific to Joy UI. We should do this:

1. list the component imports first
```js
import { Alert as JoyAlert } from '@mui/joy';
import JoyAutocomplete from '@mui/joy/Autocomplete';

// the list should be ['JoyAlert', 'JoyAutocomplete']
```

2. use `.findJSXElements($name)` and then do the transform as you already did.

Reply:
Hey, Thanks, Jun.

I changed it to the following in this [commit](https://github.com/mui/material-ui/pull/34997/commits/25e2e762be8dcf8c44ba28fc0f89298b63252334). Please let me know if I did it correctly.

```jsx
  root
    .find(j.ImportDeclaration)
    .filter(({ node }) => {
      return node.source.value.startsWith('@mui/joy');
    })
    .forEach((path) => {
      path.node.specifiers.forEach((node) => {
        root.findJSXElements(node.local.name).forEach((elementPath) => {
          if (elementPath.node.type !== 'JSXElement') {
            return;
          }

          elementPath.node.openingElement.attributes.forEach((elementNode) => {
            if (elementNode.type !== 'JSXAttribute') {
              return;
            }

            switch (elementNode.name.name) {
              case 'components':
                transformComponentsProp(elementNode);
                break;

              case 'componentsProps':
                transformComponentsPropsProp(elementNode);
                break;

              default:
            }
          });
        });
      });
    });

  const transformed = root.findJSXElements();
  ```